packages/core/src/typecheck.ts

```diff
@@ -0,0 +1,539 @@
+      )
+    );
+}
+
+/**
+ * Type guard for Row
+ * @param r variable to check
+ * @returns {boolean}
+ */
+export function isRow(r: unknown): r is Row {
+  return isSubset(r) || isAggregate(r);
+}
+
+/**
+ * Type guard for BookmarkedSelection
+ * @param b variable to check
+ * @returns {boolean}
+ */
+export function isBookmarkedSelection(b: unknown): b is BookmarkedSelection {
+  return isBookmark(b)
+  && b.type === 'elements'
+  && Object.hasOwn(b, 'selection')
+  && isElementSelection((b as BookmarkedSelection).selection);
+}
+
+/**
+ * Determines if the given object is a valid UpsetConfig using the CURRENT version.
+ * @privateRemarks
+ * This needs to be updated each time a new version is added. Since it's intended to be an exhaustive
+ * type guard against unknown types coming in, you cannot make any assumptions about the shape of the input config-
+ * exhastively check every property of new fields that you add to the config.
+ * @param config The object to check.
+ * @returns {boolean} Whether the object is a valid UpsetConfig.
+ */
+export function isUpsetConfig(config: unknown): config is UpsetConfig {
+  /* eslint-disable no-console */
+  if (!(
+    !!config
+    && typeof config === 'object'
+    && Object.hasOwn(config, 'plotInformation')
+    && Object.hasOwn(config, 'horizontal')
+    && Object.hasOwn(config, 'firstAggregateBy')
+    && Object.hasOwn(config, 'firstOverlapDegree')
+    && Object.hasOwn(config, 'secondAggregateBy')
+    && Object.hasOwn(config, 'secondOverlapDegree')
+    && Object.hasOwn(config, 'sortVisibleBy')
+    && Object.hasOwn(config, 'sortBy')
+    && Object.hasOwn(config, 'sortByOrder')
+    && Object.hasOwn(config, 'filters')
+    && Object.hasOwn(config, 'visibleSets')
+    && Object.hasOwn(config, 'visibleAttributes')
+    && Object.hasOwn(config, 'bookmarks')
+    && Object.hasOwn(config, 'collapsed')
+    && Object.hasOwn(config, 'plots')
+    && Object.hasOwn(config, 'allSets')
+    && Object.hasOwn(config, 'selected')
+    && Object.hasOwn(config, 'elementSelection')
+    && Object.hasOwn(config, 'version')
+    && Object.hasOwn(config, 'useUserAlt')
+    && Object.hasOwn(config, 'userAltText')
+  )) {
+    console.warn('Upset config is missing required fields');
+    return false;
+  }
+
+  // Put fields we've confirmed exist into vars to avoid repeating necessary casts
+  const {
+    plotInformation, horizontal, firstAggregateBy, firstOverlapDegree, secondAggregateBy, secondOverlapDegree,
+    sortVisibleBy, sortBy, sortByOrder, filters, visibleSets, visibleAttributes, attributePlots, bookmarks, collapsed,
+    plots, allSets, selected, elementSelection, version, useUserAlt, userAltText,
+  } = config as UpsetConfig;
+
+  // Check that the fields are of the correct type
+  // Start with plot info
+  if (typeof plotInformation !== 'object') {
+    console.warn('Upset config error: Plot information is not an object');
+    return false;
+  }
```

Comment:
I think that's fine; I haven't run into any scenarios where more than one field was out of compliance anyway. Since it logs to console right away it's pretty easy to update the field, re-run, confirm that field is now good, and continue.